src/osdc/ObjectCacher.h

```
@@ -8,13 +8,14 @@
 #include "include/Context.h"
 #include "include/xlist.h"
 #include "include/common_fwd.h"
+#include "include/unordered_map.h"
 
 #include "common/Cond.h"
 #include "common/Finisher.h"
+#include "common/snap_types.h" // for class SnapContext
 #include "common/Thread.h"
 #include "common/zipkin_trace.h"
 
-#include "Objecter.h"
```

I think this breaks the build:

```
/usr/bin/ar qc ../../lib/libosdc.a CMakeFiles/osdc.dir/Filer.cc.o CMakeFiles/osdc.dir/ObjectCacher.cc.o CMakeFiles/osdc.dir/error_code.cc.o
...
[ 76%] Linking CXX executable ../../../bin/ceph_test_mon_workloadgen
cd /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.3.0-7730-gf0d53703/rpm/el9/BUILD/ceph-19.3.0-7730-gf0d53703/redhat-linux-build/src/test/mon && /usr/bin/cmake -E cmake_link_script CMakeFiles/ceph_test_mon_workloadgen.dir/link.txt --verbose=1
/usr/bin/c++ -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong   -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Og -g -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -rdynamic -pie CMakeFiles/ceph_test_mon_workloadgen.dir/test_mon_workloadgen.cc.o -o ../../../bin/ceph_test_mon_workloadgen  -Wl,-rpath,/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.3.0-7730-gf0d53703/rpm/el9/BUILD/ceph-19.3.0-7730-gf0d53703/redhat-linux-build/lib: ../../../lib/libos.a ../../../lib/libosdc.a ../../../lib/libglobal.a -ldl /usr/lib64/librt.a -lresolv -ldl ../../../lib/libblk.a /lib64/libaio.so ../../../lib/libkv.a ../../../lib/libheap_profiler.a /lib64/libtcmalloc.so ../../rocksdb/librocksdb.a /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.3.0-7730-gf0d53703/rpm/el9/BUILD/ceph-19.3.0-7730-gf0d53703/src/liburing/src/liburing.a /usr/lib64/libsnappy.so /usr/lib64/liblz4.so /usr/lib64/libz.so /usr/lib64/libfuse.so ../../../lib/libceph-common.so.2 -lresolv ../../opentelemetry-cpp/sdk/src/trace/libopentelemetry_trace.a ../../opentelemetry-cpp/sdk/src/resource/libopentelemetry_resources.a ../../opentelemetry-cpp/sdk/src/common/libopentelemetry_common.a ../../opentelemetry-cpp/exporters/jaeger/libopentelemetry_exporter_jaeger_trace.a ../../opentelemetry-cpp/ext/src/http/client/curl/libopentelemetry_http_client_curl.a /usr/lib64/libcurl.so /usr/lib64/libthrift.so ../../../lib/libextblkdev.a -lcap ../../../lib/libjson_spirit.a ../../../lib/libcommon_utf8.a ../../../lib/liberasure_code.a ../../../boost/lib/libboost_thread.a ../../../boost/lib/libboost_chrono.a ../../../boost/lib/libboost_atomic.a ../../../boost/lib/libboost_system.a ../../../boost/lib/libboost_random.a ../../../boost/lib/libboost_program_options.a ../../../boost/lib/libboost_date_time.a ../../../boost/lib/libboost_iostreams.a ../../../boost/lib/libboost_regex.a /usr/lib64/libblkid.so -ldl /usr/lib64/libcrypto.so /usr/lib64/libudev.so /usr/lib64/libibverbs.so /usr/lib64/librdmacm.so /usr/lib64/libz.so  -Wl,--as-needed -latomic 
/usr/bin/ld: /tmp/ccjkdDIT.ltrans4.ltrans.o: in function `std::_Sp_counted_ptr_inplace<CrushWrapper, std::allocator<CrushWrapper>, (__gnu_cxx::_Lock_policy)2>::_M_dispose()':
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.3.0-7730-gf0d53703/rpm/el9/BUILD/ceph-19.3.0-7730-gf0d53703/src/crush/CrushWrapper.h:98: undefined reference to `crush_destroy(crush_map*)'
collect2: error: ld returned 1 exit status
make[2]: *** [src/test/mon/CMakeFiles/ceph_test_mon_workloadgen.dir/build.make:139: bin/ceph_test_mon_workloadgen] Error 1
```

Objecter.cc is part of libceph-common (yes, I know: ugh) and not part of libosdc.a. But we do actually link to libceph-common.so in this `ld` invocation. I'm not sure off-hand what the cause is. Do you see it?

---

See: https://shaman.ceph.com/builds/ceph/wip-pdonnell-testing-20250217.200327-debug/f0d537038a58ba4eee9965af406e73a87ef9a0fa/default/426536/

---

I'll try to reproduce it, but right now, I have no obvious explanation.
These are the things I mean when I say Ceph builds are very fragile - a lot is wrong and works only by chance usually - but if you stray from the standard path (standard build options), Ceph builds will suddenly fail for no obvious reason.
I want Ceph builds to be more robust, and I'll get to the root of this bug that is revealed here (not caused by this change, only revealed).

---

@batrick, this is a general Ceph bug that is just waiting to occur. Indeed I found this bug already last month and fixed it here: https://github.com/ceph/ceph/pull/60490/commits/fd77c77d322bb6e46c467e248ffc86cd0f5b24b3

This happens because `CrushWrapper.h` gets included in many many source files, and the C++ compiler generates code for the inline destructor `CrushWrapper::~CrushWrapper()` in each of these compilation units, effectively adding a dependency on `crush_destroy()` without however linking the library.

Most of the time, nothing bad happens, because this dependency is silently discarded, but depending on the include order and linker order, this dependency on `crush_destroy()` materializes in the final program and if that program is not linked with the crush library, the linker will fail.

I'll add the commit to this PR.

---

Now in a separate PR: https://github.com/ceph/ceph/pull/61872
(Should be merged before this one)

---

I'll test with this PR.

---

Thanks, I figured it must be something like that.